### PR TITLE
feat: port project to bzlmod.

### PR DESCRIPTION
**Pull request by finn-ball**
Ported to bzlmod:
https://github.com/nelhage/rules_boost/issues/546

It is now possible to depend on this project with:

```
bazel_dep(name = "com_github_nelhage_rules_boost")
```

### CODE REVIEW

**Comment by finn-ball** [comment]
@nelhage - I don't seem to be able to reproduce those CI errors on my machine.

**Comment by finn-ball** [comment]
We need to think of a better way to address this:
```
"-Iexternal/_main~non_module_dependencies~org_lzma_lzma/src/common",
```

One solution could be creating a repository rule that generates this string for you. At the moment this is very brittle and may break for projects that depend on this.

**Comment by finn-ball** [comment]
An error I get when swapping our for `bazel_dep(name = "xz", version = "5.4.5.bcr.1")`:
```
 error: undefined reference to 'lzma_stream_decoder'
```

**Comment by Attempt3035** [comment]
Had an old go at this before, ditched as I was trying to add boost to bzlmod directly but ran out of time for the project. Reinstated in #562 as it might help! I also noted build errors with lzma on my machine, so using LZMA straight from bcr in the PR, but I have no idea if that will be backwards compatible with workspace (or if bazel just lets users use both alongside, if so, this should work fine as it stands!), if you have the time available @finn-ball to test then that PR should be good to go, even though I fully believe boost build rules should be natively rewritten for BCR in the long run, [see here for work in progress](https://github.com/bazelbuild/bazel-central-registry/pull/1577)

I'm using this on a mac with bazel 7 and bzlmod which is now default. I haven't tried with workspace or on many other systems just yet, testing needed! Let me know if there's anything further I can help with!

**Comment by Attempt3035** [comment]
Ohhh I may have misunderstood your PR @finn-ball, you just upgraded the deps to BCR modules, rather than making this whole project bzlmod compatible? I like the dep upgrade, would be great if this PR could get finished off, would go nicely with #562 anyway!

**Comment by finn-ball** [comment]
I updated some dependencies to use bazel modules but also the changes were compatible for this to be used with BCR. I had some of my own internal bazel projects depend on this as a bazel module.

I couldn't reproduce the error in CI, so couldn't solve the error and no one else showed interest.

**Comment by finn-ball** [comment]
Using the `bazel_dep` of `zlib` generates this error in CI which I could not reproduce locally:
```
external/com_github_nelhage_rules_boost~~non_module_dependencies~boost/libs/iostreams/src/zlib.cpp:20:10: error: use of private header from outside its module: 'zlib.h' [-Wprivate-header]
#include "zlib.h"   // Jean-loup Gailly's and Mark Adler's "zlib.h" header.
         ^
1 error generated.
```

If this is merged, an issue needs to be created to resolve this.

**Comment by finn-ball** [comment]
@nelhage - this now passes CI. Please merge.

**Comment by Attempt3035** [comment]
Looks great, good job @finn-ball!🙌

**Comment by finn-ball** [comment]
Any thoughts @Vertexwahn ?

**Comment by Vertexwahn** [comment]
@finn-ball Once we merge this this the old approach of using `WORKSPACE `files will not working anymore - personally for me no problem ;) - @cpsauer Are you fine with moving over to Bzlmod and dropping the WORKSPACE approach?

**Comment by finn-ball** [comment]
You'll be forced to use bzlmod eventually as the `WORKSPACE` will be depreciated in the next major release iirc. With bazel 7, bzlmod is on by default.

File: boost/boost.bzl [hunk 1]
```diff
@@ -164,11 +118,10 @@ def boost_deps():
...
 
-    # We're pointing at hedronvision's mirror of google/boringssl:master-with-bazel to get Renovate auto-update. Otherwise, Renovate will keep moving us back to master, which doesn't support Bazel. See https://github.com/renovatebot/renovate/discussions/24854
-    maybe(
-        http_archive,
-        name = "openssl",
-        url = "https://github.com/hedronvision/boringssl/archive/0f1a639954dd7ab86f5f4ddd8b4e2edbea492acd.tar.gz",
-        sha256 = "7ce152bdce1b85344cc36c6b255aab36905d39187c2c2f797a69d5ad220076ee",
-        strip_prefix = "boringssl-0f1a639954dd7ab86f5f4ddd8b4e2edbea492acd",
+    http_archive(
+        name = "zlib",
```
Comment (hofbi) [inline boost/boost.bzl:122]: Why do you keep zlib here? It is available in the bazel registry so you could get it via bzlmod

Comment (finn-ball) [reply boost/boost.bzl:122]: See the above error in this PR! If this is merged, then an issue should be created to resolve this.

**Review by Vertexwahn — APPROVED** [review]

**Review by hofbi — APPROVED** [review]

**Comment by Vertexwahn** [comment]
@finn-ball Thanks for your contribution!

File: boost/boost.bzl (hunk 1 above)
Comment (finn-ball) [reply boost/boost.bzl:122]: https://github.com/nelhage/rules_boost/pull/565

See here for the currently failing build when we switch zlib to BCR

**Pull request merged**